File: Engine/script/runtimelibrary/CameraComponent.cs
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU
 
http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
using System;
using System.Runtime.CompilerServices;
using ScriptRuntime;

namespace ScriptRuntime
{
    /// <summary>
    /// 摄像机组件
    /// 指定渲染与视口相关的属性
    /// </summary>
    public partial class CameraComponent : Component
    {
        public static readonly System.Type thisType = typeof(CameraComponent);

        private CameraComponent(DummyClass__ dummy)
        {

        }
        /// <summary>
        /// 摄像机组件构造函数
        /// 默认创建一个透视投影的摄像机
        /// </summary>
        public CameraComponent()
        {
            ICall_CameraComponent_Bind(this);
            SetupPerspectiveFovRH(MathHelper.PiOver4, 1, 1000);
        }
        /// <summary>
        /// 创建正交投影
        /// </summary>
        /// <param name="height">正交投影的高度</param>
        /// <param name="zNear">指定正交投影的近裁剪面</param>
        /// <param name="zFar">指定正交投影的远裁剪面</param>
        public void SetupOrthogonal(float height, float zNear, float zFar)
        {
            ICall_CameraComponent_SetProj(this, height, zNear, zFar, Orth);
        }
        /// <summary>
        /// 创建透视投影
        /// </summary>
        /// <param name="fov">摄像机的纵向视角(角度)</param>
        /// <param name="zNear">指定透视投影的近裁剪面</param>
        /// <param name="zFar">指定透视投影的远裁剪面</param>
        public void SetupPerspectiveFovRH(float fov, float zNear, float zFar)
        {
            ICall_CameraComponent_SetProj(this, fov, zNear, zFar, Persp);
        }
        /// <summary>
        /// 获取摄像机类型
        /// </summary>
        public ViewType ViewType
        {
            get
            {
                return converViewType(ICall_CameraComponent_GetViewType(this));
            }
        }
        /// <summary>
        /// 判断摄像机是否为透视投影
        /// </summary>
        public bool IsPerspective
        {
            get
            {
                return Persp == ICall_CameraComponent_GetViewType(this);
            }
        }
        /// <summary>
        /// 获取摄像机的纵向视角(角度)
        /// </summary>
        public float Fov
        {
            get
            {
                return ICall_CameraComponent_GetFov(this);
            }
        }
        /// <summary>
        /// 获取摄像机的横纵比
        /// </summary>
        public float Aspect
        {
            get
            {
                return ICall_CameraComponent_GetAspect(this);
            }
        }
        /// <summary>
        /// 获取与设置近裁剪面的距离
        /// </summary>
        public float NearClipPlane
        {
            get
            {
                return ICall_CameraComponent_GetZNear(this);
            }
            set
            {
                if ( value < 0.01f )
                {
                    return;
                }

                ICall_CameraComponent_SetZNear(this, value);

                if ( value >= FarClipPlane )
                {
                    FarClipPlane = value + 0.01f;
                }

                if ( ShadowDistance < value )
                {
                    ShadowDistance = value;
                }

            }
        }
        /// <summary>
        /// 获取与设置远裁剪面的距离
        /// </summary>
        public float FarClipPlane
        {
            get
            {
                return ICall_CameraComponent_GetZFar(this);
            }
            set
            {
                if ( value < 0.02f )
                {
                    return;
                }

                ICall_CameraComponent_SetZFar(this, value);

                if ( NearClipPlane >= value )
                {
                    NearClipPlane = value - 0.01f;
                }

                if ( ShadowDistance > value )
                {
                    ShadowDistance = value;
                }
            }
        }
        /// <summary>
        /// 获取远裁剪面的宽度
        /// </summary>
        public float FarWidth
        {
            get
            {
                return ICall_CameraComponent_GetFarWidth(this);
            }
        }
        /// <summary>
        /// 获取远裁剪面的高度
        /// </summary>
        public float FarHeight
        {
            get
            {
                return ICall_CameraComponent_GetFarHeight(this);
            }
        }
        /// <summary>
        /// 获取近裁剪面的宽度
        /// </summary>
        public float NearWidth
        {
            get
            {
                return ICall_CameraComponent_GetNearWidth(this);
            }
        }
        /// <summary>
        /// 获取近裁剪面的高度
        /// </summary>
        public float NearHeight
        {
            get
            {
                return ICall_CameraComponent_GetNearHeight(this);
            }
        }


        /// <summary>
        /// 获取与设置阴影的距离
        /// </summary>
        public float ShadowDistance
        {
            get
            {
                return ICall_CameraComponent_GetShadowDistance(this);
            }
            set
            {
                if ( value <= FarClipPlane && value >= NearClipPlane )
                {
                    ICall_CameraComponent_SetShadowDistance(this, value);                    
                }
            }
        }


        /// <summary>
        /// 获取摄像机的相机变换矩阵
        /// </summary>
        public Matrix44 View
        {
            get
            {
                Matrix44 temp;
                ICall_CameraComponent_GetView(this, out temp);
                return temp;
            }
        }
        /// <summary>
        /// 获取摄像机的透视投影矩阵
        /// </summary>
        public Matrix44 Proj
        {
            get
            {
                Matrix44 temp;
                ICall_CameraComponent_GetProj(this, out temp);
                return temp;
            }
        }

        /// <summary>
        /// 获取摄像机的变换矩阵
        /// 透视投影矩阵乘以相机变换矩阵(OpenGL)
        /// </summary>
        public Matrix44 ViewProj
        {
            get
            {
                Matrix44 temp;
                ICall_CameraComponent_GetViewProj(this, out temp);
                return temp;
            }
        }

        /// <summary>
        /// 设置渲染到纹理
        /// </summary>
        /// <param name="rtt">指定渲染到纹理的目标</param>
        public void SetRenderTarget(RenderToTexture rtt)
        {
            if (null != rtt)
            {
                ICall_CameraComponent_SetRenderToTexture(this, rtt);
            }
        }

        /// <summary>
        /// 设置光照纹理
        /// </summary>
        /// <param name="rtt">指定光照纹理的目标</param>
        public void SetLightLitTexture(RenderToTexture rtt)
        {
            if (null != rtt)
            {
                ICall_CameraComponent_SetLightLitTexture(this, rtt);
            }
        }

        /// <summary>
        /// 为摄像机创建透视投影
        /// </summary>
        /// <param name="fov">纵向视角的角度</param>
        /// <param name="aspect">横纵比</param>
        /// <param name="zNear">近裁剪面距离</param>
        /// <param name="zFar">远裁剪面距离</param>
        public void SetupPerspectiveFovRH(float fov, float aspect, float zNear, float zFar)
        {
            ICall_CameraComponent_SetupPerspectiveFovRH(this, fov, aspect, zNear, zFar);
        }

        /// <summary>
        /// 创建一个用于水面反射与折射的投影矩阵
        /// </summary>
        /// <param name="fov">纵向视角的角度</param>
        /// <param name="aspect">横纵比</param>
        /// <param name="zNear">近裁剪面距离</param>
        /// <param name="zFar">远裁剪面距离</param>
        /// <param name="clipPlane">反射与折射的平面</param>
        /// <param name="reflect">设置是否为反射</param>
        public void SetupPerspectiveFovRHEx(float fov, float aspect, float zNear, float zFar, ref Vector4 clipPlane, bool reflect)
        {
            ICall_CameraComponent_SetupPerspectiveFovRHEx(this, fov, aspect, zNear, zFar, ref clipPlane, reflect);
        }

        /// <summary>
        /// 获取与设置是否渲染深度
        /// </summary>
        public bool RenderDepth
        {
            set
            {
                ICall_CameraComponent_SetRenderDepth(this, value);
            }
            get
            {
                return ICall_CameraComponent_GetRenderDepth(this);
            }

        }

        /// <summary>
        /// 获取与设置是否正常渲染
        /// </summary>
        public bool RenderNormal
        {
            set
            {
                ICall_CameraComponent_SetRenderNormal(this, value);
            }
            get
            {
                return ICall_CameraComponent_GetRenderNormal(this);
            }
        }

        /// <summary>
        /// 获取摄像机变换矩阵
        /// </summary>
        public Matrix44 ViewTransform
        {
            get
            {
                Matrix44 temp;
                ICall_CameraComponent_GetViewTransForm(this, out temp);
                return temp;
            }
        }
  
        /// <summary>
        /// 设置与获取用户自定义渲染
        /// </summary>
        public bool RenderCustomised
        {
            set
            {
                ICall_CameraComponent_SetCustomise(this, value);
            }
            get
            {
                return ICall_CameraComponent_IsCustomise(this);
            }
        }

        /// <summary>
        /// 获取与设置摄像机的剔除标记位
        /// </summary>
        public LayerMark CullMask
        {
            set
            {
                ICall_CameraComponent_SetCullMask(this, value.MarkAsUINT);
            }
            get
            {
                return ICall_CameraComponent_GetCullMask(this);
            }
        }

        /// <summary>
        /// 获取与设置摄像机的剔除标记位
        /// </summary>
        public uint CullMaskUINT
        {
            set
            {
                ICall_CameraComponent_SetCullMask(this, value);
            }
            get
            {
                return ICall_CameraComponent_GetCullMask(this);
            }
        }

        internal static ViewType converViewType(int viewType)
        {
            return (ViewType)viewType;
        }
        
        internal static int converViewType(ViewType viewType)
        {
            return (int)viewType;
        }
        private static int Orth = converViewType(ViewType.Orthogonal);
        private static int Persp = converViewType(ViewType.Perspective);

        /// <summary>
        /// 设置是否使用前置渲染事件
        /// </summary>
        public bool UseBeforeDrawEvent
        {
            set 
            {
                ICall_CameraComponent_SetUseBeforeDrawEvent(this, value);
            }
            get
            {
                return ICall_CameraComponent_GetUseBeforeDrawEvent(this);
            }
        }

        /// <summary>
        /// 设置是否渲染阴影
        /// </summary>
        public bool RenderShadowMap
        {
            set
            {
                ICall_CameraComponent_SetRenderShadowMap(this, value);
            }
            get
            {
                return ICall_CameraComponent_GetRenderShadowMap(this);
            }
        }
    }
}

File: Engine/app/graphicfeature/components/cameracomponent.cc
/****************************************************************************
Copyright (c) 2011-2013,WebJet Business Division,CYOU
 
http://www.genesis-3d.com.cn

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/
#include "stdneb.h"
#include "graphicfeature/components/cameracomponent.h"
#include "appframework/actor.h"
#include "appframework/component.h"
#include "graphicfeature/graphicsfeature.h"

#include "graphicfeature/graphicsfeatureprotocol.h"

#include "app/apputil/mouserayutil.h"
#include "appframework/actormanager.h"


namespace App
{
__ImplementClass(CameraComponent, 'CAMP', Component);

using namespace Math;
using namespace Graphic;

//------------------------------------------------------------------------------
/**
*/
CameraComponent::CameraComponent()
: m_bUseSkyBox(false)
, m_bAttached(false)
, m_bSkyBoxDirty(false)
, m_bInit(false)

{
	m_camera = Graphic::Camera::Create();
	m_camera->SetOwner(this);
	m_camera->SetCameraOrder(Graphic::eCO_Main);
}

//------------------------------------------------------------------------------
/**
*/
CameraComponent::~CameraComponent()
{
	m_camera = NULL;
}
//------------------------------------------------------------------------------
/**
*/
void
CameraComponent::SetupCallbacks()
{
	Super::SetupCallbacks();
	this->mActor->RegisterComponentCallback(this, BeginFrame);
	this->mActor->RegisterComponentCallback(this, OnFrame);
    this->mActor->RegisterComponentCallback(this, MoveAfter);
}

void CameraComponent::_OnBeginFrame()
{
	Super::_OnBeginFrame();
	if (m_bSkyBoxDirty)
	{
		DetachSkyBox();
		AttachSkyBox();
	}

}
//------------------------------------------------------------------------
void CameraComponent::_OnFrame()
{
	Super::_OnFrame();
}

void CameraComponent::GetReferenceResourceId(Util::Array<Resources::ReferenceResource>& list) const
{
	//这个类只有编辑器会用到，所以不用统计。
	Super::GetReferenceResourceId(list);
}
//------------------------------------------------------------------------------
/**
*/
void
CameraComponent::OnActivate()
{
	Super::OnActivate();

	n_assert( m_camera.isvalid() );
	n_assert( GraphicsFeature::HasInstance() );
	n_assert( mActor );

	if ( mActor )
	{
		m_camera->SetTransform(  mActor->GetWorldTranslateNoScale()  );
		m_camera->OnTransformChanged();
	}
	AttachCamera();
#ifndef __SCRIPT_COMMIT__
	mActor->Check_OnRenderPostEffect_Bind();
#endif
}
//------------------------------------------------------------------------------
/**
*/
void
CameraComponent::OnDeactivate()
{
	DetachCamera();
#ifndef __SCRIPT_COMMIT__
	mActor->Check_OnRenderPostEffect_Unbind();
#endif
	Super::OnDeactivate();
}
//------------------------------------------------------------------------------
/**
*/
void 
CameraComponent::OnRenderSceneChanged()
{
	if (m_camera.isvalid())
	{
		m_camera->SetRenderScene(mActor->GetRenderScene());
	}
}

//------------------------------------------------------------------------------
/**
*/
void
CameraComponent::SetupAcceptedMessages()
{   
    Super::SetupAcceptedMessages();
}

//------------------------------------------------------------------------------
/**
*/
void
CameraComponent::HandleMessage(const GPtr<Messaging::Message>& msg)
{
}
//------------------------------------------------------------------------
bool CameraComponent::GetProjParam( float* nearWidth , float* nearHeight , float* nearPlaneDis , float* farPlaneDis )
{
	*nearWidth  = m_camera->GetCameraSetting().GetNearWidth() ;
	*nearHeight = m_camera->GetCameraSetting().GetNearHeight() ;
	*nearPlaneDis = m_camera->GetCameraSetting().GetZNear() ;
	*farPlaneDis = m_camera->GetCameraSetting().GetZFar();

	return true ;
}
//------------------------------------------------------------------------
bool CameraComponent::SetProjParam( float nearW , float nearH , float nearDis , float farDis )
{

	Graphic::CameraSetting cs = m_camera->GetCameraSetting();
	
	float fovy = 2.0f * atan(nearH/(2.0f * nearDis));//

	cs.SetupPerspectiveFovRH( fovy , nearW / nearH , nearDis , farDis ) ;

	m_camera->SetCameraSetting( cs ) ;

	return true ;
}
//------------------------------------------------------------------------
bool CameraComponent::SetProj(float fovOrHeight, float zNear, float zFar, Graphic::Camera::ViewType vt)
{
	m_camera->SetProj(vt, fovOrHeight, zNear, zFar);
	return true ;
}

bool CameraComponent::SetProj(float fovOrHeight, float zNear, float zFar)
{
	return SetProj(fovOrHeight, zNear, zFar, m_camera->GetViewType());
}

void CameraComponent::SetZNear(float zNear)
{
	const CameraSetting& cs = m_camera->GetCameraSetting();
	Camera::ViewType vt = m_camera->GetViewType();
	switch (vt)
	{
	case Camera::VT_persp:
		{
			SetProj(cs.GetFov(), zNear, cs.GetZFar());
		}
		break;
	case Camera::VT_ortho:
		{
			SetProj(cs.GetNearHeight(), zNear, cs.GetZFar());
		}
		break;
	default:
		{
			n_error("No suport this Camera::ViewType: %d", (int)vt);//[zhongdaohuan]不允许有这种情况发生。
		}
		break;
	}
}

void CameraComponent::SetZFar(float zFar)
{
	const CameraSetting& cs = m_camera->GetCameraSetting();
	Camera::ViewType vt = m_camera->GetViewType();
	switch (vt)
	{
	case Camera::VT_persp:
		{
			SetProj(cs.GetFov(), cs.GetZNear(), zFar, vt);
		}
		break;
	case Camera::VT_ortho:
		{
			SetProj(cs.GetNearHeight(), cs.GetZNear(), zFar, vt);
		}
		break;
	default:
		{
			n_error("No suport this Camera::ViewType: %d", (int)vt);//[zhongdaohuan]不允许有这种情况发生。
		}
		break;
	}
}
//------------------------------------------------------------------------
void CameraComponent::_OnMoveAfter()
{
	if ( mActor )
	{
		m_camera->SetTransform( mActor->GetWorldTranslateNoScale() );
		m_camera->OnTransformChanged();		
	}

}
//------------------------------------------------------------------------
void CameraComponent::CopyFrom( const GPtr<Component>& pComponent )
{
	if( !pComponent.isvalid()  )
		return;
	if( !pComponent->GetRtti()->IsDerivedFrom( *(this->GetRtti()) ) )
		return;

	GPtr<CameraComponent> pSource = pComponent.downcast<CameraComponent>();
	GPtr<Graphic::Camera> pSourceCamera = pSource->GetCameraObject();
	n_assert( pSource->m_camera.isvalid() );

	m_camera->CopyParam(pSourceCamera);
}

//------------------------------------------------------------------------
void CameraComponent::AttachCamera()
{
	if ( m_bAttached )
	{
		return;
	}

	if (eCO_PuppetMain != m_camera->GetCameraOrder())
	{
		const GPtr<GraphicSystem>& gs = GraphicsFeature::Instance()->GetGraphicSystem();
		n_assert( gs.isvalid() );
		if ( gs.isvalid() )
		{
			m_camera->SetRenderScene(mActor->GetRenderScene());
			m_camera->Setup();
			gs->AddCamera( m_camera );

			//if (m_camera->GetCameraOrder() == eCO_Main)
			//{
			//	GraphicsFeature::Instance()->AssignDefaultCamera( mActor );
			//}
		}
	}

	m_bAttached = true;

	AttachSkyBox();
}
//------------------------------------------------------------------------
void CameraComponent::DetachCamera()
{
	if ( !m_bAttached )
	{
		return;
	}

	const GPtr<GraphicSystem>& gs = GraphicsFeature::Instance()->GetGraphicSystem();
	n_assert( gs.isvalid() );
	if ( gs.isvalid() )
	{
		m_camera->Discard();
		gs->RemoveCamera( m_camera );	
		m_camera->SetRenderScene(NULL);
	}
	
	
	m_bAttached = false;

	DetachSkyBox();
}

void CameraComponent::SetCullMarkID(CullMarkID id)
{
	//m_cullMarkID = id;
	RenderLayer rl = RenderLayerConverter::Convert(id);
	if (m_camera.isvalid())
	{
		m_camera->SetCullMask(rl);
	}
}

void CameraComponent::AttachSkyBox()
{
}

void CameraComponent::DetachSkyBox()
{
	const GPtr<GraphicSystem>& gs = GraphicsFeature::Instance()->GetGraphicSystem();
	n_assert( gs.isvalid() );
}

bool CameraComponent::IsUseSkyBox(void)
{
	return m_bUseSkyBox;
}

void CameraComponent::SetUseSkyBox(bool useSkyBox)
{
}



	void CameraComponent::SetRenderCustom( bool custom )
	{
		if (m_camera.isvalid())
		{
			m_camera->SetUseCustomMaterial(custom);
		}
	}

	bool CameraComponent::IsRenderCustom()
	{
		if (m_camera.isvalid())
		{
			return m_camera->IsRenderCustom();
		}
		return false;
	}

	void CameraComponent::SetRenderNormal( bool normal )
	{
		if (m_camera.isvalid())
		{
			m_camera->SetRenderNormal(normal);
		}
	}

	bool CameraComponent::IsRenderNormal()
	{
		if (m_camera.isvalid())
		{
			return m_camera->IsRenderNormal();
		}
		return false;
	}

	void CameraComponent::SetRenderDepth( bool RenderDepth )
	{
		if (m_camera.isvalid())
		{
			m_camera->SetRenderDepth(RenderDepth);
		}
	}

	bool CameraComponent::IsRenderDepth()
	{
		if (m_camera.isvalid())
		{
			return m_camera->HasDepthMap();
		}
		return false;
	}

	void CameraComponent::SetRenderToTexture( const GPtr<Graphic::RenderToTexture>& rtt )
	{
		if (m_camera.isvalid())
		{
			return m_camera->SetRenderToTexture(rtt);
		}
	}

	void CameraComponent::SetPerspectiveFovRH( float fov, float aspect, float zNear, float zFar )
	{

		m_camera->SetPerspectiveFovRH(fov, aspect, zNear, zFar);
	}

	void CameraComponent::SetPerspectiveFovRHEx( float fov, float aspect, float zNear, float zFar,const Math::float4& clipPlane, bool reflection )
	{
		m_camera->SetPerspectiveFovRH(fov, aspect, zNear, zFar, clipPlane, reflection);
	}

	bool
		CameraComponent::IsAllResourceLoaded()
	{
		return true;
	}

	void
		CameraComponent::SetupAllResource()
	{
		return;
	}

}; // namespace GraphicsFeature
